lib/datadog/appsec/contrib/devise/patcher.rb

```diff
@@ -0,0 +1,160 @@
+# frozen_string_literal: true
+
+require_relative '../patcher'
+require_relative 'resource'
+require_relative 'ext'
+require_relative '../../../kit/appsec/events'
+
+module Datadog
+  module AppSec
+    module Contrib
+      module Devise
+        AUTOMATED_USER_EVENT_TAGGING_BLOCK = proc do |trace, event|
+          trace.set_tag("_dd.appsec.events.#{event}.auto.mode", AppSec.settings.automated_track_user_events)
+        end
+
+        private_constant :AUTOMATED_USER_EVENT_TAGGING_BLOCK
+
+        # Hook in devise validate method
+        module AuthenticablePatch
+          # rubocop:disable Metrics/MethodLength, Metrics/PerceivedComplexity
+          def validate(resource, &block)
+            result = super
+            return result unless AppSec.enabled?
+            return result if AppSec.settings.automated_track_user_events == Ext::DISABLED_MODE
+
+            active_trace = defined?(Datadog::Tracing) && Datadog::Tracing.active_trace
+
+            return result unless active_trace
+
+            devise_resource = resource ? Resource.new(resource) : nil
+
+            event_information = {}
+            user_id = nil
+
+            if AppSec.settings.automated_track_user_events == Ext::EXTENDED_MODE && devise_resource
```

Comment:
Wondering if we should have predicate methods in there someday instead of comparing values, like `AppSec.settings.automated_track_user_events.extended?`